tool/src/org/antlr/v4/codegen/UnicodeEscapes.java

```diff
@@ -8,31 +8,42 @@
 
 /**
  * Utility class to escape Unicode code points using various
- * languages' syntaxes.
+ * languages' syntax.
  */
-public abstract class UnicodeEscapes {
-	static public void appendJavaStyleEscapedCodePoint(int codePoint, StringBuilder sb) {
-		if (Character.isSupplementaryCodePoint(codePoint)) {
-			// char is not an 'integral' type, so we have to explicitly convert
-			// to int before passing to the %X formatter or else it throws.
-			sb.append(String.format("\\u%04X", (int)Character.highSurrogate(codePoint)));
-			sb.append(String.format("\\u%04X", (int)Character.lowSurrogate(codePoint)));
-		}
-		else {
-			sb.append(String.format("\\u%04X", codePoint));
-		}
+public class UnicodeEscapes {
```

Comment:
oh I see. You have made it an enumerator type. I'm not sure I see what the advantages here and I see a disadvantage and that we now have to update switch statements for new target in a number of locations. And target type it's not the right word because we mean target language. Type sounds like data type or something.

Reply:
I'm not sure what disadvantage you mean. With `String` it will be the same `switch` statement. Also, some branches of `switch` are common for several targets.

I've refactored code, separated responsibilities. For instance, JavaScript target now does not call unrelated Java method as it [was before](https://github.com/antlr/antlr4/blob/master/tool/src/org/antlr/v4/codegen/target/JavaScriptTarget.java#L146), but has the common branch with Java instead.

Reply:
as you've discovered, a big disadvantage is no one can make their own target.